src/lib/utilities/sokoban.ts

```diff
@@ -0,0 +1,512 @@
+	 * }
+	 *
+	 * // update message
+	 * return interaction.update({
+	 *  content: updatedLevel,
+	 *	components: buildGameControls(
+	 *		encodedLevelFromLevelComponent,
+	 *    level.checkNonviableMoves(),
+	 *		startTimestamp === 0 ? Date.now() : startTimestamp,
+	 *		(moves ? Number(moves) : 0) + 1
+	 *  ),
+	 *	flags: MessageFlags.Ephemeral
+	 * });
+	 * ```
+	 *
+	 */
+	public constructor(public readonly gameComponents: EmojiGameComponent[]) {
+		const coordinatedGameComponents: GameComponent[] = [];
+
+		let maxColumnIndice = 0;
+		let maxRowIndice = 0;
+		let currentRowStartIndex = 0;
+		let x = 0;
+		let y = 0;
+		// get the max indices for rows and columns and coordinate components for simplified null fillings
+		for (let i = 0; i < gameComponents.length; i++) {
+			const component = gameComponents[i];
+			coordinatedGameComponents.push({ x, y, component });
+			if (component === EmojiGameComponent.NewLine) {
+				maxColumnIndice = Math.max(maxColumnIndice, i - currentRowStartIndex);
+				currentRowStartIndex = i + 1;
+				maxRowIndice++;
+				x = 0;
+				y++;
+				continue;
+			}
+			if (i === gameComponents.length - 1) maxColumnIndice = Math.max(maxColumnIndice, i - currentRowStartIndex);
+			x++;
+		}
+		this.rowIndices = maxRowIndice;
+		this.columnIndices = maxColumnIndice;
+
+		// fill the game with empty components to make it a perfect rectangle
+		for (let i = 0; i < coordinatedGameComponents.length; i++) {
+			if (coordinatedGameComponents[i].component === EmojiGameComponent.NewLine) {
+				const gap = maxColumnIndice - coordinatedGameComponents[i].x;
+				coordinatedGameComponents.splice(i, 0, ...Array(gap).fill({ x: 0, y: 0, component: EmojiGameComponent.Null }));
+				i += gap;
+			}
+		}
+
+		this.gameComponents = coordinatedGameComponents.map((c) => c.component);
+	}
+
+	/**
+	 * Gets a component at the given position.
+	 *
+	 * @param x x-axis position.
+	 * @param y y-axis position.
+	 * @returns the component at the given position.
+	 */
+	public get(x: number, y: number): GameComponent {
+		return { component: this.gameComponents[y * (this.columnIndices + 1) + x], x, y };
+	}
+
+	/**
+	 * Sets a component at the given position.
+	 *
+	 * @param x x-axis position.
+	 * @param y y-axis position.
+	 * @param component the component to set the position to.
+	 */
+	public set(x: number, y: number, component: EmojiGameComponent) {
+		this.gameComponents[y * (this.columnIndices + 1) + x] = component;
+	}
+
+	/**
+	 * Gets the player component.
+	 *
+	 * @returns the player component.
+	 */
+	public getPlayer(): GameComponent {
+		let x = 0;
+		let y = 0;
+		const component = this.gameComponents.find((component) => {
+			if (component === EmojiGameComponent.NewLine) {
+				y++;
+				x = 0;
+				return false;
+			}
+			x++;
+			if ([EmojiGameComponent.Player, EmojiGameComponent.PlayerTarget].includes(component)) {
+				return true;
+			}
+			return false;
+		})!;
+		return { component, x: Math.max(0, x - 1), y };
```

Comment:
As discussed in DMs, there's the reason for the code not already being the clean implementation that you've provided, which I'll look into fixing, but besides that, I don't think an error throw is a good idea considering users can import their own levels and I think it would be better if we returned a `Result.err()` with a localized message.

Reply:
But I’m not saying remove the Result, I’m saying you should use `err` rather than `Result.err`. The `err` function is available as a top export: `import { err } from '@sapphire/iterator-utilities';`

Reply:
*I think you misinterpreted my commented as being under the `Result` and `Option` specific suggestion you gave, this comment is regarding the 1D indexing for getPlayer*